### PR TITLE
force user input in autocomplete

### LINKED ISSUE #174: Autocomplete accept value that is not on the list
**I'm submitting a feature request**

* **Angular Version:**7
major.minor.patch

* **Library Version:**
major.minor.patch

* **Browser:** all

* **Language:**
TypeScript 3.4.5

**Current behavior:**
When added a value on the autocomplete that is not on the list then pressed enter the value goes back to the previous selected value.

**Expected/desired behavior:**
When added a value that is not on the list the autocomplete should give the user a list but if not found the user should be able to add a new one

* **What is the motivation / use case for changing the behavior?**
In case the user wants to add a value that is not included on the list.

PS: I am not sure if this feature already exists

### PR DESCRIPTION
**Pull request by loonix**
Fixes #174

### CODE REVIEW

**Comment by ghiscoding** [comment]
Thanks for the PR, however if this option is really only used with the Autocomplete Editor, I would rather move it into the `editorOptions`. 

I try to keep, as much as possible, only options that can be used by all Editors inside the `columnEditor.interface.ts`, else the `editorOptions` is a better fit even though that property is generic.

**Comment by loonix** [comment]
That sounds reasonable, I changed it to the `editorOptions` and took it out of the `columnEditor`. I removed the type not sure if you want to add it on the `multipleSelectOption.interface.ts`

**Comment by ghiscoding** [comment]
If you want the intellisense to work, could maybe change the `editorOptions` to the following
```ts
editorOptions?: MultipleSelectOption | forceUserInput?: boolean | any;
```

I know that keeping the `any` as a type is not too good, it kind of undo the intellisense feature, maybe someday I should add all possible options for each Editors but it's kinda long to do that. For example the  Date Editor uses [Flatpickr](https://flatpickr.js.org/examples/) and so I would have to go and all every possible options into a separate interface and then modify the `editorOptions` like so
```ts
editorOptions?: MultipleSelectOption | FlatpickrOptions | AutoCompleteOptions;
```
but that is kinda long, I don't have time to do that. Unless you or anyone else have a better way of doing it with TypeScript

**Comment by loonix** [comment]
I am happy to give a hand on that, and that would be a lot better since we can track the options straight away (maybe on a next pull request)

**Comment by ghiscoding** [comment]
Sure, I created the [version 2.x - Project](https://github.com/ghiscoding/Angular-Slickgrid/projects/1) and added some tasks there. I didn't use much before, so if you have better ways of displaying that, let me know, I think it should be enough to start. I can reference a PR to the task when it opens up

File: src/app/modules/angular-slickgrid/editors/autoCompleteEditor.ts [hunk 1]
```diff
@@ -69,7 +71,7 @@ export class AutoCompleteEditor implements Editor {
     const title = this.columnEditor && this.columnEditor.title || '';
     this.labelName = this.customStructure && this.customStructure.label || 'label';
     this.valueName = this.customStructure && this.customStructure.value || 'value';
-
+    this.forceUserInput = this.columnEditor && this.columnEditor.editorOptions && this.columnEditor.editorOptions.forceUserInput ? this.columnEditor.editorOptions.forceUserInput : false;
```
Comment (ghiscoding) [inline src/app/modules/angular-slickgrid/editors/autoCompleteEditor.ts]: To simplify this line, you could maybe use a Getter, for example see this [Getter](https://github.com/ghiscoding/Angular-Slickgrid/blob/c4a29417c01de8fdf81a999afa25a517a6c1e5ce/src/app/modules/angular-slickgrid/editors/autoCompleteEditor.ts#L50). It's just a suggestion, I could also accept the code without it too

Comment (loonix) [reply src/app/modules/angular-slickgrid/editors/autoCompleteEditor.ts]: Done, makes sense to refactor it  :)

Comment (ghiscoding) [reply src/app/modules/angular-slickgrid/editors/autoCompleteEditor.ts]: Oh man I'm really sorry, I was not clear in my comment, can you change it to the following
```ts
get editorOptions() {
    return this.columnEditor && this.columnEditor.editorOptions || {};
  }
```
then you can call your boolean values (and any other editor options) like so
```ts
serializeValue() {
  if (this.editorOptions.forceUserInput) {
...
```

File: src/app/modules/angular-slickgrid/editors/autoCompleteEditor.ts [hunk 2]
```diff
@@ -63,13 +65,17 @@ export class AutoCompleteEditor implements Editor {
...
+  }
+
   init(): void {
     const columnId = this.columnDef && this.columnDef.id;
     const placeholder = this.columnEditor && this.columnEditor.placeholder || '';
     const title = this.columnEditor && this.columnEditor.title || '';
     this.labelName = this.customStructure && this.customStructure.label || 'label';
     this.valueName = this.customStructure && this.customStructure.value || 'value';
-
+    this.forceUserInput = this.forceUserInputChecker;
```
Comment (ghiscoding) [inline src/app/modules/angular-slickgrid/editors/autoCompleteEditor.ts]: do you really need this line after using the Getter?

Comment (loonix) [reply src/app/modules/angular-slickgrid/editors/autoCompleteEditor.ts]: not after the changes you requested, removed it.

File: src/app/modules/angular-slickgrid/editors/autoCompleteEditor.ts (hunk 1 above)
Comment (loonix) [reply src/app/modules/angular-slickgrid/editors/autoCompleteEditor.ts]: No problem at all (never used getter before) made a new pr, let me know if it is what you have in mind. Thanks

**Comment by ghiscoding** [comment]
Thanks looks good now :)

**Comment by ghiscoding** [comment]
Will merge now, I was trying to fix the new Circle-CI and Codecov for unit tests coverage. I finally found the correct config. Sweet, I now have full CI (continuous integration) in the project :)

**Comment by ghiscoding** [comment]
@loonix 
Sorry for the delay, as you have noticed I wanted to wait for the other PR to be merged. 
Your update is now available under latest version [2.6.0](https://github.com/ghiscoding/Angular-Slickgrid/releases/tag/v2.6.0) on [NPM](https://www.npmjs.com/package/angular-slickgrid)

**Comment by loonix** [comment]
No problem, many thanks for telling me. Is there a slack chanel or irc chanel you use to go?

**Comment by ghiscoding** [comment]
I don't really use Slack and there's no channel or any kind since I'm alone on this lol

**Pull request merged**